src/Runtime/Distributions/Automata/ListAutomaton.cs

```diff
@@ -122,28 +122,27 @@ public ListAutomaton()
             var elementStateWeightSum = Weight.Zero;
             foreach (var element in transitionElements)
             {
-                Weight prevStateWeight;
-                if (!elementStatesWeights.TryGetWeight(element.destIndex, out prevStateWeight))
+                if (!elementStatesWeights.TryGetWeight(element.destIndex, out var prevStateWeight))
                 {
                     prevStateWeight = Weight.Zero;
                 }
 
-                elementStatesWeights[element.destIndex] = Weight.Sum(prevStateWeight, element.weight);
-                elementStateWeightSum = Weight.Sum(elementStateWeightSum, element.weight);
+                elementStatesWeights[element.destIndex] = prevStateWeight + element.weight;
+                elementStateWeightSum += element.weight;
             }
 
             var destinationState = new Determinization.WeightedStateSet();
             foreach (KeyValuePair<int, Weight> stateIdWithWeight in elementStatesWeights)
             {
                 if (stateIdWithWeight.Value.LogValue > LogEps)
                 {
-                    Weight stateWeight = Weight.Product(stateIdWithWeight.Value, Weight.Inverse(elementStateWeightSum));
+                    Weight stateWeight = stateIdWithWeight.Value * Weight.Inverse(elementStateWeightSum);
                     destinationState.Add(stateIdWithWeight.Key, stateWeight);
                 }
             }
 
-            Weight transitionWeight = Weight.Product(Weight.FromValue(1), elementStateWeightSum);
```

Comment:
This seems strange in the original code

Reply:
Same for me. But original code mimics the StringAutomaton. Where instead of Weight.FromValue(1) the Weight.FromValue(segmentLength) is used. I simplified this bit in rewritten code.